packages/actor-query-operation-group/lib/GroupsState.ts

```diff
@@ -0,0 +1,149 @@
+      const aggregators: { [key: string]: AggregateEvaluator } = {};
+      for (const i in this.pattern.aggregates) {
+        const aggregate = this.pattern.aggregates[i];
+        const key = termToString(aggregate.variable);
+        aggregators[key] = new AggregateEvaluator(aggregate, bindings);
+      }
+
+      const group: IGroup = { aggregators, bindings: grouper };
+      this.groups.set(groupHash, group);
+
+      if (this.distinctHashes) {
+        const bindingsHash = this.hashBindings(bindings);
+        this.distinctHashes.set(groupHash, new Set([bindingsHash]));
+      }
+
+    } else {
+      // Group already exists
+      // Update all the aggregators with the input binding
+      const group = this.groups.get(groupHash);
+      for (const i in this.pattern.aggregates) {
+        const aggregate = this.pattern.aggregates[i];
+
+        // If distinct, check first wether we have inserted these values already
+        if (aggregate.distinct) {
+          const hash = this.hashBindings(bindings);
+          if (this.distinctHashes!.get(groupHash).has(hash)) {
+            continue;
+          } else {
+            this.distinctHashes!.get(groupHash).add(hash);
+          }
+        }
+
+        const variable = termToString(aggregate.variable);
+        group.aggregators[variable].put(bindings);
+      }
+    }
+  }
+
+  /**
+   * Collect the result of the current state. This returns a Bindings per group,
+   * and a (possibly empty) Bindings in case the no Bindings have been consumed yet.
+   */
+  public collectResults(): Bindings[] {
+    // Collect groups
+    let rows: Bindings[] = Array.from(this.groups, ([_, group]) => {
+      const { bindings: groupBindings, aggregators } = group;
+
+      // Collect aggregator bindings
+      // If the aggregate errorred, the result will be undefined
+      const aggBindings: { [key: string]: Term } = {};
+      for (const variable in aggregators) {
+        const value = aggregators[variable].result();
+        if (value !== undefined) { // Filter undefined
+          aggBindings[variable] = value;
+        }
+      }
+
+      // Merge grouping bindings and aggregator bindings
+      return groupBindings.merge(aggBindings);
+    });
+
+    // Case: No Input
+    // Some aggregators still define an output on the empty input
+    // Result is a single Bindings
+    if (rows.length === 0) {
+      const single: { [key: string]: Term } = {};
+      for (const i in this.pattern.aggregates) {
+        const aggregate = this.pattern.aggregates[i];
+        const key = termToString(aggregate.variable);
+        const value = AggregateEvaluator.emptyValue(aggregate);
+        if (value !== undefined) {
+          single[key] = value;
+        }
+      }
+      rows = [Bindings(single)];
+    }
+
+    return rows;
+  }
+
+  /**
+   * @param {Bindings} bindings - Bindings to hash
+   */
+  private hashBindings(bindings: Bindings): BindingsHash {
```

Comment:
I would suggest depending on `@comunica/actor-abstract-bindings-hash` and using the `hash` function from there. Feel free to make adjustments to that package if needed.

(Now that I look at that package, I see some badly abstracted code, so feel free to improve that as you see fit) (Also, that package seems to be missing `json-stable-stringify` in its dependency list)

Reply:
Should abstract bindings hash become a peer or regular dependency?

Reply:
A regular dependency.

Reply:
I refrained from refactoring abstract bindings hash too much since a refactor in line with how it's currently used would require more work that is not related to this PR and affect other packages.

Reply:
I am gonna revert the name changes even, as they are API incompatible, and this might cause issues.